csharp/src/Apache.Arrow/Arrays/ArrayData.cs

```diff
@@ -28,12 +27,25 @@ public sealed class ArrayData : IDisposable
 
         public readonly IArrowType DataType;
         public readonly int Length;
-        public readonly int NullCount;
+        private int _nullCount;
         public readonly int Offset;
         public readonly ArrowBuffer[] Buffers;
         public readonly ArrayData[] Children;
         public readonly ArrayData Dictionary; // Only used for dictionary type
 
+        public int NullCount
+        {
+            get
+            {
+                if (_nullCount == RecalculateNullCount)
+                {
+                    _nullCount = ComputeNullCount();
```

Comment:
This might result in duplicate work if multiple threads try to access the null count concurrently, but I don't imagine this is a big concern, and the C++ implementation has the same behaviour.